src/combinatorics/set_family/lym.lean

```diff
@@ -0,0 +1,272 @@
+Prove upward local LYM.
+
+Provide equality cases. Local LYM gives that the equality case of LYM and Sperner is precisely when
+`𝒜` is a middle layer.
+
+Most of the machinery (`from_above`, `from_below` and `falling`) is useful more generally in grade
+orders.
+
+## References
+
+* http://b-mehta.github.io/maths-notes/iii/mich/combinatorics.pdf
+* http://discretemath.imp.fu-berlin.de/DMII-2015-16/kruskal.pdf
+
+## Tags
+
+shadow, lym, slice, sperner, antichain
+-/
+
+open finset nat
+open_locale big_operators finset_family
+
+variables {𝕜 α : Type*} [linear_ordered_field 𝕜]
+
+-- generalize `tsub_le_tsub_left` to `preorder`
+-- generalize `tsub_le_iff_left` to `add_comm_semigroup`
+
+lemma tsub_tsub_le_tsub_add [preorder α] [add_comm_monoid α] [has_sub α] [has_ordered_sub α]
+  [covariant_class α α (+) (≤)] {a b c : α} :
+  a - (b - c) ≤ a - b + c :=
+tsub_le_iff_right.2 $ calc
+    a ≤ a - b + b : le_tsub_add
+  ... ≤ a - b + (c + (b - c)) : add_le_add_left le_add_tsub _
+  ... = a - b + c + (b - c) : (add_assoc _ _ _).symm
+
+-- lemma tsub_tsub_le_tsub_add' [preorder α] [add_comm_monoid α] [has_sub α] [has_ordered_sub α]
+--   [covariant_class α α (+) (≤)] {a b c : α} :
+--   a - (b - c) ≤ a - b + c :=
+-- by { rw [←tsub_le_iff_right], have := tsub_tsub,
+--     sorry,
+--  exact tsub_le_tsub_left le_tsub_add _ }
+
+namespace finset
+
+/-- The only element of `insert a s` that is not an element of `s` is `a`. -/
+lemma eq_of_not_mem_of_mem_insert [decidable_eq α] {a b : α} {s : finset α} (hb : b ∉ s)
+  (ha : b ∈ insert a s) :
+  b = a :=
+(mem_insert.1 ha).resolve_right hb
+
+lemma insert_inj [decidable_eq α] {a b : α} {s : finset α} (ha : a ∉ s) :
+  insert a s = insert b s ↔ a = b :=
+begin
+  refine ⟨λ h, eq_of_not_mem_of_mem_insert ha _, congr_arg _⟩,
+  rw ←h,
+  exact mem_insert_self _ _,
+end
+
+lemma insert_inj_on' [decidable_eq α] (s : finset α) : set.inj_on (λ a, insert a s) sᶜ :=
+λ a ha b _, (insert_inj ha).1
+
+lemma insert_inj_on [decidable_eq α] [fintype α] (s : finset α) :
+  set.inj_on (λ a, insert a s) (sᶜ : finset α) :=
+by { rw coe_compl, exact s.insert_inj_on' }
+
+@[simp]
+lemma card_erase_of_mem' [decidable_eq α] {a : α} {s : finset α} (ha : a ∈ s) :
+  (s.erase a).card = s.card - 1 :=
+card_erase_of_mem ha
+
+lemma sdiff_nonempty [decidable_eq α] {s t : finset α} : (s \ t).nonempty ↔ ¬ s ⊆ t :=
+by rw [nonempty_iff_ne_empty, ne.def, sdiff_eq_empty_iff_subset]
+
+lemma exists_eq_insert_iff [decidable_eq α] {s t : finset α} :
+  (∃ a ∉ s, insert a s = t) ↔ s ⊆ t ∧ s.card + 1 = t.card :=
+begin
+  refine ⟨_, _⟩,
+  { rintro ⟨a, ha, rfl⟩,
+    exact ⟨subset_insert _ _, (card_insert_of_not_mem ha).symm⟩ },
+  { rintro ⟨hst, h⟩,
+    obtain ⟨a, ha⟩ : ∃ a, t \ s = {a},
+    { exact card_eq_one.1 (by rw [card_sdiff hst, ←h, add_tsub_cancel_left]) },
+    refine ⟨a, λ hs, (_ : a ∉ {a}) $ mem_singleton_self _,
+      by rw [insert_eq, ←ha, sdiff_union_of_subset hst]⟩,
+    rw ←ha,
+    exact not_mem_sdiff_of_mem_right hs }
+end
```

Comment:
You're planning to move everything from here upwards into a more appropriate file, right?